### Airbnb Partie 1 - Solutionnaire

Nom : Alla Iakymenko

Matricule : 11308140 

## Importation
Utilisez cette section pour importer la ou les librairies pertinentes et lire le fichier.

In [57]:
import pandas as pd
import numpy as np
Airbnb b= pd.read_csv("./fichiers/Airbnb.csv", sep = ",")
Airbnb


ID                                                Nom  \
0          2539                 Clean & quiet apt home by the park   
1          2595                              Skylit Midtown Castle   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !   
3          3831                    Cozy Entire Floor of Brownstone   
4          5022   Entire Apt: Spacious Studio/Loft by central park   
...         ...                                                ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   
48891  36485057      Affordable room in Bushwick/East Williamsburg   
48892  36485431            Sunny Studio at Historical Neighborhood   
48893  36485609               43rd St. Time Square-cozy single bed   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen   

       ID proprio    Nom proprio Arrondissement            Quartier  Latitude  \
0            2787           John       Brooklyn          Kensington  40.64749   
1            2845       Jennifer      Manhattan             Midtown  40.75362   
2            4632      Elisabeth      Manhattan              Harlem  40.80902   
3            4869    LisaRoxanne       Brooklyn        Clinton Hill  40.68514   
4            7192          Laura      Manhattan         East Harlem  40.79851   
...           ...            ...            ...                 ...       ...   
48890     8232441        Sabrina       Brooklyn  Bedford-Stuyvesant  40.67853   
48891     6570630        Marisol       Brooklyn            Bushwick  40.70184   
48892    23492952  Ilgar & Aysel      Manhattan              Harlem  40.81475   
48893    30985759            Taz      Manhattan      Hell's Kitchen  40.75751   
48894    68119814     Christophe      Manhattan      Hell's Kitchen  40.76404   

       Longitude             Type  Prix  Minimum nuits  Nombre avis  \
0      -73.97237     Private room   149            1.0            9   
1      -73.98377  Entire home/apt   225            1.0           45   
2      -73.94190     Private room   150            3.0            0   
3      -73.95976  Entire home/apt    89            1.0          270   
4      -73.94399  Entire home/apt    80           10.0            9   
...          ...              ...   ...            ...          ...   
48890  -73.94995     Private room    70            2.0            0   
48891  -73.93317     Private room    40            4.0            0   
48892  -73.94867  Entire home/apt   115           10.0            0   
48893  -73.99112      Shared room    55            1.0            0   
48894  -73.98933     Private room    90            7.0            0   

      Dernier avis  Avis par mois  Affichages proprio  \
0       2021-10-19           0.21                 6.0   
1       2022-05-21           0.38                 2.0   
2              NaN            NaN                 1.0   
3       2022-07-05           4.64                 1.0   
4       2021-11-19           0.10                 1.0   
...            ...            ...                 ...   
48890          NaN            NaN                 2.0   
48891          NaN            NaN                 2.0   
48892          NaN            NaN                 1.0   
48893          NaN            NaN                 6.0   
48894          NaN            NaN                 1.0   

       Jours disponibles annuel  
0                         365.0  
1                         355.0  
2                         365.0  
3                         194.0  
4                           0.0  
...                         ...  
48890                       9.0  
48891                      36.0  
48892                      27.0  
48893                       2.0  
48894                      23.0  

[48895 rows x 16 columns]

In [71]:
Airbnb["ID proprio"].unique()

array([     2787,      2845,      4632, ..., 274321313,  23492952,
        68119814], dtype=int64)

### Question 1 : Quel est l'ID proprio de la 500e location affichée dans la liste?
Par exemple, l'ID proprio de la 1ere location affichée dans la liste est 2787.

In [58]:
location500e = Airbnb.iloc[499,2 ]
location500e

#Reponse:  l'ID proprio de la 500e location affichée dans la liste est 846309.

846309

### Question 2 : Quel est le nom de propriétaire de location qui apparaît le plus souvent?

In [59]:
Question2 = Airbnb.pivot_table(values = "ID", index = "Nom proprio", aggfunc = "count", fill_value = 0 )
Question2.sort_values(["ID"],ascending=False).iloc[0]
#Le nom de propriétaire de location qui apparaît le plus souvent - Michael.

ID    417
Name: Michael, dtype: int64

### Question 3 : Quel est le prix médian des locations à Manhattan?

In [60]:
Question3 = Airbnb.loc[Airbnb["Arrondissement"] == "Manhattan",["Prix","Arrondissement"]]["Prix"].median()
Question3
#Reponse: le prix médian des locations à Manhattan est 150.

150.0

### Question 4 : Combien de locations nécessitent une réservation minimum d'exactement 365 nuits?

In [61]:
Question4 =(Airbnb.loc[Airbnb["Minimum nuits"]== 365, :])["ID"].count()
Question4
#Reponse: 29 de locations nécessitent une réservation minimum d'exactement 365 nuits.

29

### Question 5 : Combien y a-t-il de locations où le nom du propriétaire est manquant?

In [62]:
Question5 = Airbnb[Airbnb["Nom proprio"].isnull()]["ID"].count()
Question5
#Reponse: 21 de locations où le nom du propriétaire est manquant.

21

### Question 6 : Dans quel arrondissement le quartier avec le plus grand nombre d'avis se trouve-t-il?

In [67]:
Airbnb.groupby("Type").mean()["Prix"]#["Private room"]

Type
Entire home/apt    211.785654
Private room        89.816115
Shared room         70.412740
Name: Prix, dtype: float64

In [ ]:
#metod 1
Airbnb.groupby(["Arrondissement", "Quartier"]).sum().sort_values("Nombre avis", ascending = False)["Nombre avis"].iloc[0:].index[0][0]

In [ ]:
#metod 2
test6 = Airbnb.pivot_table(values = "Nombre avis", index = ["Arrondissement","Quartier"],aggfunc = "sum", fill_value = 0).sort_values("Nombre avis", ascending = False)
arrondissement = test6.iloc[0].name[0]
arrondissement

In [ ]:
#metod3 
Question6 = Airbnb.groupby(["Arrondissement", "Quartier"]).sum()["Nombre avis"]
Question6

In [ ]:
#metod3 continue
Question6[Question6 == Question6.max()]

### Question 7 : Quel est le prix moyen par type de résidence?

In [ ]:
Question7 = Airbnb.pivot_table(values ="Prix", index = "Type", aggfunc ="mean", fill_value = 0)
Question7

### Question 8 : Quels sont les 3 quartiers les moins chers en moyenne?

In [ ]:
Question8 = Airbnb.pivot_table(values = "Prix", index = "Quartier", aggfunc = "mean").sort_values("Prix").iloc[0:3, 0]
Question8

### Question 9 : Quel est le prix moyen des locations ayant reçu plus de 100 avis et se trouvant à Brooklyn?

In [ ]:
#correct
Question9 = Airbnb[(Airbnb["Arrondissement"] == "Brooklyn") & (Airbnb["Nombre avis"] >= 100)]["Prix"].mean()
Question9 = round(Question9,2)
print("Le prix moyen des locations ayant reçu plus de 100 avis et se trouvant à Brooklyn est {}.".format(Question9))

### Question 10 : Quelle proportion de l'ensemble des locations provient de la combinaison des 10 propriétaires ayant le plus grand nombre de locations?
Cette question signifie que la réponse doit être une seule valeur et non pas dix valeurs.

In [ ]:
Question10 = Airbnb.pivot_table(values = "ID", index ="ID proprio", aggfunc = "count")
Question10

In [ ]:
proprio_10 = Question10.sort_values("ID", ascending = False).iloc[0:10].sum()
proprio_10

In [ ]:
proprio_total= Question10.sort_values("ID").sum()
proprio_total

In [ ]:
proprio_10_proportion = proprio_10/proprio_total*100
proprio_10_proportion = round(proprio_10_proportion[0],2)
print("La proportion de l'ensemble des locations provient de la combinaison des 10 propriétaires ayant le plus grand nombre de locations est {}%.".format(proprio_10_proportion)) 

### Question 11 : Dans quel arrondissement trouve-t-on le plus haut taux de locations du type "Shared room"? Quel est ce taux?

In [ ]:
Question11 = Airbnb.pivot_table(values = "Prix", index = "Type", columns = "Arrondissement", aggfunc = "count")
Question11

In [ ]:
# Calculer la proportion des locations de type "Shared room" et les trier par ordre décroissant
taux_de_locations = Question11.loc["Shared room"] / Question11.sum(axis=0)
taux_de_loc_shared_room = taux_de_locations.sort_values(ascending=False)
# Trouver l'arrondissement avec le taux le plus élevé de locations de type "Shared room"
arrondissement_max = taux_de_loc_shared_room.idxmax()

print("L'arrondissement avec le taux le plus élevé de locations de type 'Shared room' est l'arrondissement {}.".format(arrondissement_max))
# Récupérer la proportion la plus élevée de locations de type "Shared room"
proportion_la_plus_elevee = taux_de_loc_shared_room.iloc[0]

print("La proportion la plus élevée de locations de type 'Shared room' est de {:.2%}".format(proportion_la_plus_elevee))

### Question 12 : Deux colonnes en particulier contiennent exactement 10 052  valeurs manquantes. Quelle est la raison concrète qui explique ces valeurs manquantes? Démontrez votre affirmation.

In [ ]:
Airbnb["Dernier avis"].isnull().sum()

In [ ]:
# Afficher le nombre total de valeurs manquantes dans chaque colonne
print(Airbnb.isnull().sum())

In [ ]:
# Créer un nouveau DataFrame contenant uniquement les locations avec 0 commentaires
locations_sans_avis = Airbnb[Airbnb["Nombre avis"] == 0]

# Afficher le nombre de valeurs manquantes dans les colonnes de ce nouveau DataFrame
print(locations_sans_avis.isnull().sum())

### Question 13 : Quel est le prix moyen des locations par arrondissement et par catégorie de nombre d'avis?
On considère les catégories de nombre d'avis suivantes :
- Catégorie 1 : une location avec moins de 100 avis
- Catégorie 2 : une location avec 100 à 200 avis
- Catégorie 3 : une location avec plus de 200 avis

In [ ]:
Airbnb.loc[Airbnb["Nombre avis"]< 100, "Catégorie"] = "Catégorie 1"
Airbnb.loc[(Airbnb["Nombre avis"]>= 100) & (Airbnb["Nombre avis"]<=200), "Catégorie"] = "Catégorie 2"
Airbnb.loc[Airbnb["Nombre avis"]> 200, "Catégorie"] = "Catégorie 3"

In [ ]:
Airbnb.pivot_table(values = "Prix", index = "Arrondissement", columns = "Catégorie", aggfunc = "mean")

### Question 14 : Quel arrondissement possède la plus forte proportion de locations étant affichées depuis moins d'un an ?

In [ ]:
# Convertir la colonne de date d'affichage en datetime

Airbnb["Dernier avis"] = pd.to_datetime(Airbnb["Dernier avis"])

In [ ]:
#Pour comprendre le type de données
Airbnb.dtypes

In [ ]:
import numpy as np
import pandas as pd

# Calculer la date d'un an avant aujourd'hui
date_un_an_avant = pd.Timestamp.now() - pd.Timedelta(days=365)

# Sélectionner les locations affichées depuis moins d'un an
locations_recentes = Airbnb[Airbnb["Dernier avis"] > date_un_an_avant]

# Compter le nombre de locations récentes par arrondissement
locations_recentes_par_arrondissement = locations_recentes.groupby("Arrondissement").size()

# Calculer la proportion de locations récentes par arrondissement
proportion_locations_recentes_par_arrondissement = locations_recentes_par_arrondissement / Airbnb.groupby("Arrondissement").size()

# Trouver l'arrondissement avec la plus forte proportion de locations récentes
arrondissement_plus_fort = proportion_locations_recentes_par_arrondissement.sort_values(ascending = False).idxmax()

# Afficher le résultat
print("L'arrondissement avec la plus forte proportion de locations récentes étant affichées depuis moins d'un an est :   ", arrondissement_plus_fort)

In [ ]:
pd.Timestamp.now().date()

In [ ]:
proportion_locations_recentes_par_arrondissement

In [ ]:
df_sansna = Airbnb.dropna().copy()
df_sansna["Mois"] = df_sansna["Nombre avis"]/df_sansna["Avis par mois"]
df_sansna[df_sansna["Mois"]<12].groupby("Arrondissement").count()["ID"]

In [ ]:
df_sansna.groupby("Arrondissement").count()["ID"]

In [ ]:
df_sansna["Mois"].value_counts()

In [ ]:
Q_14 = df_sansna[df_sansna["Mois"]<12].groupby("Arrondissement").count()["ID"]/df_sansna.groupby("Arrondissement").count()["ID"]
Q_14[1]
Q_14[Q_14 == 0.2941137478456847]

### Question 15 : Quels sont les 5 mots différents qui se retrouvent le plus souvent dans les noms des locations?
La réponse ne devrait pas dépendre de l'utilisation des lettres majuscules ou minuscules. Par exemple, les mots "Location" et "location" sont les mêmes.

In [ ]:
locations = list(Airbnb["Nom"].fillna("").tolist())
locations
type(locations)

# Supposons que la liste de noms de location est stockée dans la variable `lieux`
# Définir une liste vide pour stocker tous les mots individuels
mots = []

# Collecter tous les mots de chaque nom de lieu
for location in locations:
    mots += location.lower().split()

# Compter le nombre d'occurrences de chaque mot
occurrences = {}
for mot in mots:
    if mot in occurrences:
        occurrences[mot] += 1
    else:
        occurrences[mot] = 1

# Trier les mots par ordre décroissant de nombre d'occurrences
mots_tries = sorted(occurrences, key=occurrences.get, reverse=True)

# Sélectionner les 5 premiers mots de la liste triée
cinq_mots = mots_tries[:5]

# Afficher les résultats

print("Les 5 mots les plus courants dans les noms de lieu sont :")
for mot in cinq_mots:
    print(f"{mot} ({occurrences[mot]} occurrences)")